### ISSUE TITLE
[tests][tvOS] Some System.IO tests are failing

### ISSUE THREAD

**Issue by simonrozsival** [dotnet/runtime#67853]
OS: tvOS
Architecture: arm64
Configuration: release
Runtime: mono

It appears that some System.IO tests on tvOS have been silently failing for at least a month. The #67746 bugfix removes a crash that prevented xharness from reporting the test failures properly (e.g. [rolling build from 4/10](https://dev.azure.com/dnceng/public/_build/results?buildId=1709824&view=logs&j=f4520fb1-1559-5885-1d9c-3cb3f6a85e23&t=6c7a8cfe-f92e-569a-eef9-b2ad3e13056d) with this [System.IO.Tests output](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-release-70-preview1b4d075be3db04238bd/System.IO.Tests/1/net.dot.System.IO.Tests.log?sv=2019-07-07&se=2022-04-30T18%3A39%3A43Z&sr=c&sp=rl&sig=LeGYoJpKoDnBw%2BFDvgi1JfOz98Hij1%2BoOBXDzePP4J0%3D)).

The following tests are failing:
- `DevicesPipesAndSockets.NamedPipe_ReadWrite`
- `DevicesPipesAndSockets.NamedPipe_ReadWrite_Async`
- `NamedPipeFileStreamConnectedConformanceTests.Dispose_ClosesStream`
- `NamedPipeFileStreamConnectedConformanceTests.Read_Eof_Returns0`
- `NamedPipeFileStreamConnectedConformanceTests.Disposed_ThrowsObjectDisposedException`
- `NamedPipeFileStreamConnectedConformanceTests.ZeroByteWrite_OtherDataReceivedSuccessfully`
- `NamedPipeFileStreamConnectedConformanceTests.Read_DataStoredAtDesiredOffset`
- `NamedPipeFileStreamConnectedConformanceTests.ReadAsync_ContinuesOnCurrentTaskSchedulerIfDesired`
- `NamedPipeFileStreamConnectedConformanceTests.ReadAsync_CancelPendingTask_ThrowsCancellationException`
- `NamedPipeFileStreamConnectedConformanceTests.Write_DataReadFromDesiredOffset`
- `NamedPipeFileStreamConnectedConformanceTests.ReadAsync_ContinuesOnCurrentSynchronizationContextIfDesired`
- `NamedPipeFileStreamConnectedConformanceTests.ArgumentValidation_ThrowsExpectedException`
- `NamedPipeFileStreamConnectedConformanceTests.ReadTimeout_Expires_Throws`
- `NamedPipeFileStreamConnectedConformanceTests.Flush_ValidOnWriteableStreamWithNoData_Success`
- `NamedPipeFileStreamConnectedConformanceTests.ZeroByteRead_BlocksUntilDataAvailableOrNops`
- `NamedPipeFileStreamConnectedConformanceTests.ReadAsync_CancelPendingValueTask_ThrowsCancellationException`
- `NamedPipeFileStreamConnectedConformanceTests.ReadWriteAsync_PrecanceledOperations_ThrowsCancellationException`
- `NamedPipeFileStreamConnectedConformanceTests.Timeout_Roundtrips`
- `NamedPipeFileStreamConnectedConformanceTests.ReadWrite_CustomMemoryManager_Success`
- `NamedPipeFileStreamConnectedConformanceTests.Flush_ValidOnReadableStream_Success`
- `NamedPipeFileStreamConnectedConformanceTests.ReadWriteByte_Success`
- `NamedPipeFileStreamConnectedConformanceTests.ReadWrite_MessagesSmallerThanReadBuffer_Success`
- `NamedPipeFileStreamConnectedConformanceTests.CopyToAsync_AllDataCopied`
- `NamedPipeFileStreamConnectedConformanceTests.ReadWrite_Success`
- `Directory_ExistsAsFile.FalseForNonRegularFile`
- `DirectoryInfo_EnumFSI_str_str.SearchPatternCaseSensitive`
- `FileInfo_Exists.TrueForNonRegularFile`
- `Directory_Delete_str_bool.ExtendedDirectoryWithSubdirectories`
- `Directory_Delete_str_bool.LongPathExtendedDirectory`
- `File_GetSetAttributesCommon.SettingInvalidAttributes_UnixExceptOSXAndFreeBSD(attributes: Hidden)`
- `Directory_CreateDirectory.DriveLetter_Unix`
- `Directory_EnumFSE_str_str_so.SearchPatternCaseSensitive`
- `File_ReadWriteAllBytes.ReadAllBytes_NonSeekableFileStream_InUnix`
- `FileInfo_GetSetAttributesCommon.SettingInvalidAttributes_UnixExceptOSXAndFreeBSD(attributes: Hidden)`
- `PathFile_Exists.TrueForNonRegularFile`
- `DirectoryInfo_Delete_bool.ExtendedDirectoryWithSubdirectories`
- `DirectoryInfo_Delete_bool.LongPathExtendedDirectory`
- `File_ReadWriteAllBytesAsync.ReadAllBytesAsync_NonSeekableFileStream_InUnix`
- `File_ExistsAsDirectory.FalseForNonRegularFile`
- `Directory_GetLogicalDrives.GetsValidDriveStrings_Unix`
- `DirectoryInfo_Create.DriveLetter_Unix`
- `DirectoryInfo_Exists.FalseForNonRegularFile`
- `DirectoryInfo_Delete.ExtendedDirectoryWithSubdirectories`
- `DirectoryInfo_Delete.LongPathExtendedDirectory`
- `Directory_Delete_str.ExtendedDirectoryWithSubdirectories`
- `Directory_Delete_str.LongPathExtendedDirectory`

**Comment by simonrozsival** [comment]
There is some overlap with #51371

**Comment by steveisok** [comment]
A lot of the failures are asserting at

https://github.com/dotnet/runtime/blob/51d11ebbaff4e967652e61b2b371e0d2f04c6fba/src/libraries/Common/tests/TestUtilities/System/IO/FileCleanupTestBase.cs#L159

https://github.com/dotnet/runtime/pull/63877 was intended to fix these issues and it appears they did on iOSSimulator, but we got a false positive from CI when it came to devices.

**Comment by mkhamoyan** [comment]
Closing as duplicate of #51390
After #51390 is fixed , will enable all these tests

**Comment by directhex** [comment]
Reopening. This issue is a collection of entirely unrelated issues, and likely not even tvOS specific. We have dozens of ActiveIssue for this bug number, will close when those are handled.

**Comment by directhex** [comment]
This is actually a mix of three distinct bugs:

* Tests failing which rely on Named Pipes or Unix Domain Sockets
* Tests failing which assume `cwd` is writeable
* Tests failing which assume the filesystem case sensitivity behaviour, but are wrong (this is especially annoying, as tvOS is inconsistent with tvOSSimulator)

**Comment by directhex** [comment]
Okay then. Detailed failures info:

```
// unexpected behaviour with mkfifo
System.IO.Tests.PathFile_Exists.TrueForNonRegularFile
System.IO.Tests.FileInfo_Exists.TrueForNonRegularFile
System.IO.Tests.DirectoryInfo_Exists.FalseForNonRegularFile
System.IO.Tests.File_ReadWriteAllBytesAsync.ReadAllBytesAsync_NonSeekableFileStream_InUnix
System.IO.Tests.File_ExistsAsDirectory.FalseForNonRegularFile
System.IO.Tests.Directory_ExistsAsFile.FalseForNonRegularFile
System.IO.Tests.File_ReadWriteAllBytes.ReadAllBytes_NonSeekableFileStream_InUnix
--->
Assert.Equal() Failure
Expected: 0
Actual:   -1

// assumes writable cwd
System.IO.Tests.Directory_CreateDirectory.DriveLetter_Unix
System.IO.Tests.DirectoryInfo_Create.DriveLetter_Unix
System.IO.Tests.CreateDirectoryWithUnixFileMode.DriveLetter_Unix
--->
System.UnauthorizedAccessException : Access to the path '/private/var/containers/Bundle/Application/4DE3DC40-5DBA-4461-B77E-737E989BF256/System.IO.FileSystem.Tests.app/C:' is denied.
---- System.IO.IOException : Operation not permitted

// trying to use Windows UNC from cwd
System.IO.Tests.DirectoryInfo_Delete.ExtendedDirectoryWithSubdirectories
System.IO.Tests.DirectoryInfo_Delete.LongPathExtendedDirectory
System.IO.Tests.DirectoryInfo_Delete_bool.ExtendedDirectoryWithSubdirectories
System.IO.Tests.DirectoryInfo_Delete_bool.LongPathExtendedDirectory
System.IO.Tests.Directory_Delete_str.ExtendedDirectoryWithSubdirectories
System.IO.Tests.Directory_Delete_str.LongPathExtendedDirectory
System.IO.Tests.Directory_Delete_str_bool.ExtendedDirectoryWithSubdirectories
System.IO.Tests.Directory_Delete_str_bool.LongPathExtendedDirectory
--->
System.UnauthorizedAccessException : Access to the path '/private/var/containers/Bundle/Application/4DE3DC40-5DBA-4461-B77E-737E989BF256/System.IO.FileSystem.Tests.app/\\\\?\\' is denied.
---- System.IO.IOException : Operation not permitted
```

**Comment by directhex** [comment]
Gotta use remote executor to use SetCurrentDirectory to fix the UNC & drive letter tests, apparently.
